src/java.base/share/classes/java/lang/reflect/code/bytecode/LocalsTypeMapper.java

```diff
@@ -44,54 +47,126 @@
 
 import static java.lang.classfile.attribute.StackMapFrameInfo.SimpleVerificationTypeInfo.*;
 import static java.lang.constant.ConstantDescs.*;
-import java.lang.constant.DirectMethodHandleDesc;
-import java.lang.constant.DynamicConstantDesc;
-import java.lang.constant.MethodTypeDesc;
+import java.lang.reflect.code.Value;
+import java.lang.reflect.code.type.JavaType;
+import java.util.ArrayDeque;
+import java.util.LinkedHashSet;
+import java.util.Set;
 
 final class LocalsTypeMapper {
 
-    record Frame(List<ClassDesc> stack, List<ClassDesc> locals) {}
+    private record Link(Slot slot, Link other) {}
```

Comment:
Consider enclosing the declaration of `Link` in `Slot` (with the same access control as `Slot`).